### PR TITLE
UMA authorization

### PR DESCRIPTION
**Pull request by SlevinBE**
UMA (User Managed Access) is an extension of the OAuth 2 protocol, which allows a client to request access to a protected resource, of which the permissions are strictly defined in a resource server.

The implementation in this forward-auth proxy has the client request access to all resources on the resource server (configured in Keycloak, where the Keycloak client acts and is configured like a resource server). When the authorization server denies the request for access, it returns a 403, otherwise a 200 response.
```
curl -X POST \
  http://${host}:${port}/auth/realms/${realm}/protocol/openid-connect/token \
  -H "Authorization: Bearer ${access_token}" \
  --data "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" \
  --data "audience={resource_server_client_id}"
```
More info: 
- [Keycloak UMA Support](https://www.keycloak.org/docs/4.8/authorization_services/index.html#_service_obtaining_permissions)
- [UMA 2.0 Specification](https://docs.kantarainitiative.org/uma/wg/oauth-uma-grant-2.0-05.html)
- [UMA 2 in Action](https://wso2.com/articles/2019/3/user-managed-access-in-action-part-one)

### CODE REVIEW

File: forwardauth.go [hunk 1]
```diff
@@ -164,6 +165,32 @@ func (f *ForwardAuth) ExchangeCode(r *http.Request, code string) (string, error)
 	return token.Token, err
 }
 
+// Verifies whether access is allowed to all resources of the client using the UMA protocol.
+// https://www.keycloak.org/docs/4.8/authorization_services/index.html#_service_obtaining_permissions
+
```
Comment (tsmethurst) [inline forwardauth.go]: You can make this godoc compliant with a few little changes eg:
```
// VerifyAccess checks whether access is allowed to all resources of the client using the UMA protocol.
// https://www.keycloak.org/docs/4.8/authorization_services/index.html#_service_obtaining_permissions
func (f *ForwardAuth) VerifyAccess(token string) (bool, error) {
```

File: forwardauth.go [hunk 2]
```diff
@@ -164,6 +165,32 @@ func (f *ForwardAuth) ExchangeCode(r *http.Request, code string) (string, error)
...
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		},
+	}
+
+	data := url.Values{}
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:uma-ticket")
+	data.Set("audience", fw.ClientId)
+
+	req, err := http.NewRequest(http.MethodPost, fw.TokenURL.String(), strings.NewReader(data.Encode()))
```
Comment (tsmethurst) [inline forwardauth.go]: There's no point encoding the data twice, instead you can do something like this:
```
func (f *ForwardAuth) VerifyAccess(token string) (bool, error) {
	// allow self-signed certs
	client := http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
	}

	data := url.Values{}
	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:uma-ticket")
	data.Set("audience", fw.ClientId)
	encoded := data.Encode()

	req, err := http.NewRequest(http.MethodPost, fw.TokenURL.String(), strings.NewReader(encoded))
	req.Header.Add("Authorization", "Bearer "+token)
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Content-Length", fmt.Sprintf("%d", len(encoded)))
	res, err := client.Do(req)

	// status code is 403 when not allowed by authorization server
	isAllowedAccess := err == nil && res.StatusCode == 200
	return isAllowedAccess, err
}
```

Comment (tsmethurst) [inline forwardauth.go]: There's no point in encoding the data twice, you can do something like this instead:

```
func (f *ForwardAuth) VerifyAccess(token string) (bool, error) {
	// allow self-signed certs
	client := http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
	}

	data := url.Values{}
	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:uma-ticket")
	data.Set("audience", fw.ClientId)
	encoded := data.Encode()

	req, err := http.NewRequest(http.MethodPost, fw.TokenURL.String(), strings.NewReader(encoded))
	req.Header.Add("Authorization", "Bearer "+token)
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Content-Length", fmt.Sprintf("%d", len(encoded)))
	res, err := client.Do(req)

	// status code is 403 when not allowed by authorization server
	isAllowedAccess := err == nil && res.StatusCode == 200
	return isAllowedAccess, err
}
```

**Comment by SlevinBE** [comment]
yep, tested it. Works  :)

File: forwardauth.go [hunk 3]
```diff
@@ -164,6 +165,35 @@ func (f *ForwardAuth) ExchangeCode(r *http.Request, code string) (string, error)
 	return token.Token, err
 }
 
+// VerifyAccess checks whether access is allowed to all resources of the client using the UMA protocol.
+// https://www.keycloak.org/docs/4.8/authorization_services/index.html#_service_obtaining_permissions
+func (f *ForwardAuth) VerifyAccess(token string) (bool, error) {
+	// allow self-signed certs
+	client := http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
[the remark this answers is not in the document]
Comment (SlevinBE) [reply forwardauth.go]: I added a insecureCertificates flag, which is by default false.

**Pull request merged**